### PR TITLE
Allow a ConnectionFactory (eg SslConnectionFactory) to automatically add a Customizer

### LINKED ISSUE #4814: Allow a ConnectionFactory (eg SslConnectionFactory) to automatically add a Customizer
Allows a ConnectionFactory to inspect Customizers and add optionally add a missing one.

### PR DESCRIPTION
**Pull request by gregw**
This PR for #4814 adds a `Connector.Configuring` interface to allow the `SslConnectionFactory` to ensure there is a `SecureRequestCustomizer`.

I have made this a draft as I'm a bit worried about modifying shared `HttpConfiguration` instances... not sure if it is a concern and if it is, what to do about it?

There is also a good optimisation of the request attributes for SSL (probably should be a different PR... maybe even back ported to 9.4).

### CODE REVIEW

**Comment by gregw** [comment]
@lachlan-roberts look at what I've done with `Attributes` in this PR.  This is kind of what I'm thinking we need to do with Async dispatch as well.

**Comment by joakime** [comment]
What happens when you mix `SecureRequestCustomizer` and `ForwardedRequestCustomizer` in the same connector/configuration?
Eg: Who wins the scheme and isSecure setting? (or will one undo the other?)
Does order of those customizers matter?

**Comment by gregw** [comment]
Spun out `Attributes` improvements to #4816.  Edit: I'll remove them from this PR shortly.

**Comment by gregw** [comment]
@joakime I don't think it matters, but probably should be reviewed carefully.   `SecureRequestCustomer` will setSecure if the connection is secure (eg SSL), whilst `ForwardedRequestCustomizer` will setSecure if the headers say it is secure.  I don't think either will clear secure, so the only clash is if they both trigger, but then they both should have the same action, so I think that is OK (but kind of strange setup).

**Comment by gregw** [comment]
I have forced pushed this PR with only the `ConnectionFactory.Configuring` changes.  The `Attributes` improvements are now all in #4488 

I still have a bit of a concern how this approach might affect any shared HttpConfiguration instances?

File: jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionFactory.java [hunk 1]
```diff
@@ -125,4 +125,16 @@
          */
         Detection detect(ByteBuffer buffer);
     }
+
+    /**
+     * A ConnectionFactory that can configure the connector.
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionFactory.java:130]: Please fix the javadoc, as `Configuring` is not a `ConnectionFactory`. 
Unless, only a `ConnectionFactory` can implement `Configuring`.
If something else can, then the javadoc needs to be fixed, something like:  "Implementing this class allows to configure the connector (or its component tree) at connector startup".

**Review by sbordet — APPROVED** [review]
Looks mostly ok, but there is no test that calls `setEnsureSecureRequestCustomizer()`. The implementation is quite trivial though, so maybe there is no need to.
Please update the javadocs before merge.

**Review by sbordet — CHANGES_REQUESTED** [review]
I meant to request the javadocs changes and a clarification of who can implement `Configuring`.

**Comment by sbordet** [comment]
@gregw this is a _draft_ PR - still needs to be draft?

**Comment by gregw** [comment]
@sbordet It is draft as I wanted to get feedback on the approach with regards to adding a customizer to a potentially shared HttpConfiguration.

**Comment by sbordet** [comment]
@gregw the `HttpConfiguration` may be shared among HTTP connection factories on the same connector, i.e. HTTP/1.1 and HTTP/2.
I think that is good.

However, the PR also handles the case where HTTP/1.1 and HTTP/2 have been configured with two different `HttpConfiguration`s, for example you want different buffer sizes or such.

If you want to manually configure the `SecureRequestCustomizer`, you can too.

So I think we're good.

File: jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionFactory.java (hunk 1 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionFactory.java:130]: I'll make it extend ConnectionFactory like the other 'ing' inner interfaces

**Review by sbordet — APPROVED** [review]

**Pull request merged**